instrumentation/aws-sdk/aws-sdk-2.2/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/awssdk/v2_2/AwsSdkInstrumentationModule.java

```diff
@@ -10,13 +10,49 @@
 import io.opentelemetry.javaagent.extension.instrumentation.HelperResourceBuilder;
 import io.opentelemetry.javaagent.extension.instrumentation.InstrumentationModule;
 import io.opentelemetry.javaagent.extension.instrumentation.TypeTransformer;
+import io.opentelemetry.javaagent.extension.instrumentation.internal.injection.ClassInjector;
+import io.opentelemetry.javaagent.extension.instrumentation.internal.injection.ExtendedInstrumentationModule;
+import io.opentelemetry.javaagent.extension.instrumentation.internal.injection.InjectionMode;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
 
 @AutoService(InstrumentationModule.class)
-public class AwsSdkInstrumentationModule extends AbstractAwsSdkInstrumentationModule {
+public class AwsSdkInstrumentationModule extends AbstractAwsSdkInstrumentationModule
+    implements ExtendedInstrumentationModule {
   public AwsSdkInstrumentationModule() {
     super("aws-sdk-2.2-core");
   }
 
+  @Override
+  @SuppressWarnings("unchecked")
+  public List<String> getAdditionalHelperClassNames() {
+    if (isIndyModule()) {
+      // With the invokedynamic approach, the SnsInstrumentationModule and SqsInstrumentationModule
```

Comment:
I don't think this is entirely correct. The reason why there are multiple modules is that muzzle check is done per module. Having sqs and sns instrumentation allows running muzzle checks for these separately so that the main instrumentation module can still apply even if sqs and sns dependencies are missing. Why this works for you when you dump these into one module is because you have disabled the muzzle checks for the indy instrumentation modules, if you'd add back the muzzle check code then I suspect doing this would cause muzzle failures. I think that using aws instrumentation to test the proxy code was an unfortunate complication because the multimodule thing going on there complicates things. These 3 modules are currently interacting so that if the muzzle checks for the optional modules pass they'll inject the classes that the main module will dynamically discover them. With indy these 3 module are isolated from each other so the trick of dynamically discovering classes injected by another module won't work. The easiest way around this is probably to have one instrumentation class loader per application class loader.

Reply:
Or you could choose some other instrumentation besides aws for testing the proxy.

Reply:
> These 3 modules are currently interacting so that if the muzzle checks for the optional modules pass they'll inject the classes that the main module will dynamically discover them. With indy these 3 module are isolated from each other so the trick of dynamically discovering classes injected by another module won't work. 

I see, thanks for explaining this!
My prior understanding was that the AWS instrumentation is simply about injecting the library instrumentation as is and using the SQS / SNS discovery built into that library instrumentation. However, for the agent instrumentation we also want to protect agains the case of an SQS / SNS version mismatch, correct?

> The easiest way around this is probably to have one instrumentation class loader per application class loader.

Sounds good, but I'd propose that the default should be that each `InstrumentationModule` gets its own classloader. I would expect that in most cases, especially with external extensions, this isolation is a feature, not a bug. This allows extensions to freely use utility libraries (e.g. apache commons, guava) without having to fear collisions.

I'd propose for `InstrumentationModules` to optionally declare a "module group name". Modules with the same group name instrumenting the same app-classloader would get put into the same `InstrumentationModuleClassLoader`. What do you think about this approach?

If you agree, I'd shortly outline this as a TODO comment in the AWS module so that we don't forget about.

> Or you could choose some other instrumentation besides aws for testing the proxy.

Agree, I'll revert the AWS changes and choose a different instrumentation to include in this PR.

Reply:
> Agree, I'll revert the AWS changes and choose a different instrumentation to include in this PR.
 
 So I had a look at both the apache-dubbo and azur instrumentation and noticed that their tests don't fail currently with `testIndy` even though they should? I tried messing with the instrumentations (e.g. changing the resource paths) and the tests still were green. I'm thinking that something must be wrong with the tests there?
 
 I decided to use the log4j instrumentation in this PR, because that one actually fails without the proxying with indy enabled.

Reply:
at least with dubbo the issue is that the javaagent instrumentation has `implementation(project(":instrumentation:apache-dubbo-2.7:library-autoconfigure"))` which leaks over into tests so the tests will pass even without the javaagent instrumentation because they just pick up the manual instrumentation library that gets automatically configured

Reply:
I created a PR to fix dubbo tests, azure tests (ok I ran only azure-core-1.14) fail for me, perhaps you didn't notice that they have
```java
  @Override
  public boolean isIndyModule() {
    return false;
  }
```

Reply:
> ok I ran only azure-core-1.14

How did you run those, I feel like I'm missing something.

What I did was:
 * Remove the `isIndyModule()` override from the 1.14 `AzureSdkInstrumentationModule`
 * Run the 1.14 `AzureSdkTest` from the IDE with `-PtestIndy=true` and verified that the module is actually loaded as IndyModule -> the test still passes
 
Btw what is the correct way of running selected tests from the command line?
I tried `./gradlew :instrumentation:azure-core:azure-core-1.14:javaagent:test -PtestIndy=true -i` but that seems to just pick up recent test result from some kind of cache, even if I execute a `clean` beforehand:

```
> Task :instrumentation:azure-core:azure-core-1.14:javaagent:test FROM-CACHE
Custom actions are attached to task ':instrumentation:azure-core:azure-core-1.14:javaagent:test'.
Build cache key for task ':instrumentation:azure-core:azure-core-1.14:javaagent:test' is fd9f198ac1e6354f6d8e4f302611bd17
Task ':instrumentation:azure-core:azure-core-1.14:javaagent:test' is not up-to-date because:
  Output property 'binaryResultsDirectory' file /Users/jonas/git/otel/opentelemetry-java-instrumentation/instrumentation/azure-core/azure-core-1.14/javaagent/build/test-results/test/binary has been removed.
  Output property 'binaryResultsDirectory' file /Users/jonas/git/otel/opentelemetry-java-instrumentation/instrumentation/azure-core/azure-core-1.14/javaagent/build/test-results/test/binary/output.bin has been removed.
  Output property 'binaryResultsDirectory' file /Users/jonas/git/otel/opentelemetry-java-instrumentation/instrumentation/azure-core/azure-core-1.14/javaagent/build/test-results/test/binary/output.bin.idx has been removed.
  Output property 'binaryResultsDirectory' file /Users/jonas/git/otel/opentelemetry-java-instrumentation/instrumentation/azure-core/azure-core-1.14/javaagent/build/test-results/test/binary/results.bin has been removed.
  Output property 'reports.enabledReports.html.outputLocation' file /Users/jonas/git/otel/opentelemetry-java-instrumentation/instrumentation/azure-core/azure-core-1.14/javaagent/build/reports/tests/test has been removed.
Loaded cache entry for task ':instrumentation:azure-core:azure-core-1.14:javaagent:test' with cache key fd9f198ac1e6354f6d8e4f302611bd17
Could not execute [report metric STATISTICS_COLLECT_METRICS_OVERHEAD]
Could not execute [report metric STATISTICS_COLLECT_METRICS_OVERHEAD]
Could not execute [report metric STATISTICS_COLLECT_METRICS_OVERHEAD]
Could not execute [report metric STATISTICS_COLLECT_METRICS_OVERHEAD]
Could not execute [report metric STATISTICS_COLLECT_METRICS_OVERHEAD]
Could not execute [report metric STATISTICS_COLLECT_METRICS_OVERHEAD]

```

Reply:
From command line I think you have to at least do `:instrumentation:azure-core:azure-core-1.14:javaagent:cleanTest` before and then run `:instrumentation:azure-core:azure-core-1.14:javaagent:test --no-build-cache`
I retested and it still fails for me as expected when running from ide. I also removed the `isIndyModule` method from `AzureSdkInstrumentationModule` and added `-PtestIndy=true` to the run options in idea. I ran the test on main not this branch if it matters.

Reply:
> I ran the test on main not this branch if it matters.

Looks like we have a winner there, that was the difference to make them fail for me too.

I think this can be explained by `InstrumentationModule.registerHelperResources` being non-functional for indy-modules on main. In this PR that functionally has been added back.

So for this PR the azure tests should fail because the classes referenced by the injected resources are not supposed to be present in the instrumented classloader, but apparently they are.
I suspect that this then is a testing classpath issue similar to dubbo.

This shouldn't block us on this PR. I'll try to verify my hypothesis and try to open a PR doing the same thing for azure what you did for dubbo.